### PR TITLE
Features/project task assign

### PR DESCRIPTION
**Pull request by mabdullahkhalil**

### CODE REVIEW

File: app/controllers/projects_controller.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    end
+    
+  end
+
+  def edit
+  end
+
+  def destroy
+    @project.destroy
+    redirect_to projects_path, :notice => "Deleted Successfully"
```
Comment (mujeeb-arshad) [inline app/controllers/projects_controller.rb]: follow new convention i.e notice: "Deleted Successfully
"

File: app/controllers/projects_controller.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def update
+    if @project.update(projects_params)
+      redirect_to projects_path, :notice => "Updated Successfully"
+    else
+      redirect_to projects_path, :notice => "Cannot be updated"
```
Comment (mujeeb-arshad) [inline app/controllers/projects_controller.rb]: same change

File: app/controllers/projects_controller.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+  private
+
+  def projects_params
+    params.require(:project).
+    permit(:description, :name, :start_date, :expected_end_date, :department_id)
+  end
+
+  def set_department
+    # current_employee = Employee.find(1)
```
Comment (mujeeb-arshad) [inline app/controllers/projects_controller.rb]: remove comments

File: app/controllers/projects_controller.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    @departments = current_employee.company.departments
+  end
+
+  def set_project
+    # current_employee = Employee.find(1)
```
Comment (mujeeb-arshad) [inline app/controllers/projects_controller.rb]: remove

File: app/views/tasks/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        <% elsif task.assignable_type == "Team" %>  
+        <td><%= task.assignable.name %></td>
+        <% end %>
+
+        <td><span class="badge badge-success"><%= task.status %></span></td>
+
+
+        <td>
+
+
```
Comment (mujeeb-arshad) [inline app/views/tasks/index.html.erb]: remove blank spaces

File: app/views/tasks/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        <td><span class="badge badge-success"><%= task.status %></span></td>
+
+
+        <td>
+
+       
+        
+
+
+          <%= link_to 'Edit or Assign the task', edit_a_task_page_path(:project_id => task.project.id, :task_id => task.id), class: "btn btn-outline-secondary" %>
```
Comment (mujeeb-arshad) [inline app/views/tasks/index.html.erb]: follow new convention

File: config/routes.rb [hunk 7]
```diff
@@ -1,14 +1,91 @@
 Rails.application.routes.draw do
+
+    # get 'projects/resources'
+
+
+
+
+  # The priority is based upon order of creation: first created -> highest priority.
+  # See how all your routes lay out with "rake routes".
+
```
Comment (mujeeb-arshad) [inline config/routes.rb]: remove all commented code

File: spec/models/task_spec.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    it "should raise error of invalid date start date cannot be later than end date" do 
+      expect{task2}.to raise_error(ActiveRecord::RecordInvalid)
+    end
+    
+    it "should raise error of invalid date start date should be greater than today's date" do 
+      expect{task3}.to raise_error(ActiveRecord::RecordInvalid)
+    end
+
+    it "should set the status equal to new" do 
+      binding.pry
```
Comment (mujeeb-arshad) [inline spec/models/task_spec.rb]: remove binding.pry

File: spec/rails_helper.rb [hunk 9]
```diff
@@ -20,6 +22,7 @@
 # directory. Alternatively, in the individual `*_spec.rb` files, manually
 # require only the support files necessary.
 #
```
Comment (mujeeb-arshad) [inline spec/rails_helper.rb]: remove commented code

File: test/controllers/employees_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class EmployeesControllerTest < ActionController::TestCase
+  # test "the truth" do
+  #   assert true
```
Comment (mujeeb-arshad) [inline test/controllers/employees_controller_test.rb]: remove commented code

File: test/controllers/projects_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class ProjectsControllerTest < ActionController::TestCase
+  # test "the truth" do
+  #   assert true
```
Comment (mujeeb-arshad) [inline test/controllers/projects_controller_test.rb]: remove commented code

File: test/controllers/tasks_controller_test.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class TasksControllerTest < ActionController::TestCase
+  # test "the truth" do
+  #   assert true
```
Comment (mujeeb-arshad) [inline test/controllers/tasks_controller_test.rb]: remove

File: test/controllers/teams_controller_test.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class TeamsControllerTest < ActionController::TestCase
+  # test "the truth" do
+  #   assert true
```
Comment (mujeeb-arshad) [inline test/controllers/teams_controller_test.rb]: remove

File: test/fixtures/tasks.yml [hunk 14]
```diff
@@ -0,0 +1,19 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (mujeeb-arshad) [inline test/fixtures/tasks.yml]: remove these

File: test/models/company_test.rb [hunk 15]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class CompanyTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
```
Comment (mujeeb-arshad) [inline test/models/company_test.rb]: remove

File: test/models/department_test.rb [hunk 16]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class DepartmentTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
```
Comment (mujeeb-arshad) [inline test/models/department_test.rb]: remove

File: test/models/employee_test.rb [hunk 17]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class EmployeeTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
```
Comment (mujeeb-arshad) [inline test/models/employee_test.rb]: remove

File: test/models/project_test.rb [hunk 18]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class ProjectTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
```
Comment (mujeeb-arshad) [inline test/models/project_test.rb]: remove

File: test/models/task_test.rb [hunk 19]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class TaskTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
```
Comment (mujeeb-arshad) [inline test/models/task_test.rb]: remove

File: test/models/team_test.rb [hunk 20]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class TeamTest < ActiveSupport::TestCase
+  # test "the truth" do
```
Comment (mujeeb-arshad) [inline test/models/team_test.rb]: remove

File: app/assets/javascripts/tasks.js.coffee [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+        preventDuplicates: true,
+        hintText: 'Add Team',
+        tokenLimit: 1,
+        tokenValue: 'id',
+        searchingText: 'searching teams...'}
+
+
+jQuery ->
+    $(document).ready ->
+      $('.assign_team').tokenInput('/teamslist.json', hash_new);
```
Comment (zainsaleem4338) [inline app/assets/javascripts/tasks.js.coffee]: js-assign-team

File: app/assets/javascripts/tasks.js.coffee [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+      $('.assign_employee').tokenInput('/employeeslist.json', hash_new_team);
```
Comment (zainsaleem4338) [inline app/assets/javascripts/tasks.js.coffee]: js-assign-employee

File: config/database.yml [hunk 23]
```diff
@@ -1,22 +1,27 @@
...
+#   gem install sqlite3
+#
+#   Ensure the SQLite 3 gem is defined in your Gemfile
+#   gem 'sqlite3'
+#
 default: &default
   adapter: mysql2
   encoding: utf8
-  pool: 5
+  database: ez_employee_tracking_development_2
```
Comment (zainsaleem4338) [inline config/database.yml]: why are you changing database?

File: spec/rails_helper.rb [hunk 24]
```diff
@@ -20,6 +22,7 @@
...
+# Dir[Rails.root.join('spec', 'support', '**', '*.rb')].each { |f| require f }
```
Comment (zainsaleem4338) [inline spec/rails_helper.rb]: remove

File: app/controllers/employees_controller.rb [hunk 25]
```diff
@@ -41,6 +42,14 @@ def destroy
     end
   end
 
+
+  def employees_list
+    @employees = Employee.order(:email)
```
Comment (zainsaleem4338) [inline app/controllers/employees_controller.rb]: load through company

File: app/controllers/employees_controller.rb [hunk 26]
```diff
@@ -41,6 +42,14 @@ def destroy
...
+    respond_to do |format|
+      format.json { render json: Employee.where('role != ? AND name like ?', 'admin',"%#{params[:q]}%" ) }
```
Comment (zainsaleem4338) [inline app/controllers/employees_controller.rb]: use 'admin' constant
I think it is already defined in Employee.rb

File: app/controllers/tasks_controller.rb [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  def new
+    @task = current_employee.company.projects.find(params[:id]).tasks.new
+  end
+
+  def create
+    @task = Task.new(tasks_create_params)
+    @task.set_status
+    @task.set_assignable(params[:assignable_team_id], params[:assignable_employee_id], params[:task][:assignable_type])
```
Comment (zainsaleem4338) [inline app/controllers/tasks_controller.rb]: check if params[:task] is present

File: app/models/task.rb [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+    errors.add(:expected_end_date,'cannot be before start_time') if (expected_end_date.to_date < start_date.to_date)
+  end
+
+  def check_start_date
+    errors.add(:start_date,"should be greater than equal to today's date.") if start_date.to_date < Date.today
+  end
+
+  def set_status
+    return self.status = "assigned" if !self.assignable_id.nil?
+    self.status = "new"
```
Comment (zainsaleem4338) [inline app/models/task.rb]: "new" constant

File: app/models/task.rb [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+  def check_start_and_end_date
+    errors.add(:expected_end_date,'cannot be before start_time') if (expected_end_date.to_date < start_date.to_date)
+  end
+
+  def check_start_date
+    errors.add(:start_date,"should be greater than equal to today's date.") if start_date.to_date < Date.today
+  end
+
+  def set_status
+    return self.status = "assigned" if !self.assignable_id.nil?
```
Comment (zainsaleem4338) [inline app/models/task.rb]: "assigned" constant
self.assignable.present? instead of !self.assignable_id.nil?

File: app/models/task.rb [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+    self.status = "new"
+    self.assignable_type = nil
+  end
+
+  def set_assignable(assignable_team_id, assignable_employee_id,assignable_type)
+    if !assignable_team_id.empty?
```
Comment (zainsaleem4338) [inline app/models/task.rb]: assignable_team_id.present?

File: app/models/task.rb [hunk 31]
```diff
@@ -0,0 +1,49 @@
...
+    end
+  end
+
+  def update_assignable(assignable_team_id, assignable_employee_id,assignable_type)
+    if (self.assignable_type == Task::EMPLOYEE && assignable_type == Task::EMPLOYEE && assignable_employee_id == self.assignable_id) ||
+      (self.assignable_type == Task::TEAM && assignable_type == Task::TEAM && assignable_employee_id == self.assignable_id)
+      return
+    else 
+      if assignable_type == Task::EMPLOYEE
+        self.update(assignable_type: assignable_type, assignable_id: assignable_employee_id, status: "assigned") if !assignable_employee_id.empty?
```
Comment (zainsaleem4338) [inline app/models/task.rb]: assignable_employee_id.present?

File: app/models/task.rb [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+        self.update(assignable_type: nil, assignable_id: nil, status: "new") if assignable_employee_id.empty?
```
Comment (zainsaleem4338) [inline app/models/task.rb]: assignable_employee_id.blank?

File: app/views/projects/index.html.erb [hunk 33]
```diff
@@ -0,0 +1,29 @@
...
+</div>
+<h1>Listing Projects</h1>
+<%= link_to 'New Project', new_project_path, class: "btn btn-primary" %>
+<% if !@projects.nil?%>								
+	<div class="row">
+		<% @projects.each do |project| %>
+		<tbody>
+			<div class="card col-sm-4">
+				<div class="card-body">
+					<h5 class="card-title"><%= project.name %></h5>
```
Comment (zainsaleem4338) [inline app/views/projects/index.html.erb]: spacing issues

File: app/views/tasks/_form.html.erb [hunk 34]
```diff
@@ -0,0 +1,130 @@
...
+
+    <label class="col"><input type="radio" name="adding"; value="1" class="js-toggle-team-employee-input-token team_radio" data-type="Employee">Add a task for Employee</label>
+    <label class="col"><input type="radio" name="adding"; value="2" checked="checked" class="js-toggle-team-employee-input-token"  data-type="Team"/>Add a task for Team</label>
+
+    <div class= "adding_employee_field collapse js-toggleable-input-token">
+      <%= form.label :assignable_id, "Employee Email" %><br>
+      <%= text_field_tag :assignable_employee_id, @task.assignable, class: "js-assign-employee" %>
+      <% if @task.assignable_type.nil? %>
+      <% assignable_employee_hash = nil%>
+      <% elsif @task.assignable_type == "Employee" %>
```
Comment (zainsaleem4338) [inline app/views/tasks/_form.html.erb]: constant

File: app/views/tasks/_form.html.erb [hunk 35]
```diff
@@ -0,0 +1,130 @@
...
+      <% assignable_employee_hash = [@task.assignable].to_json.html_safe %>
+      <% end %>
+      <%= javascript_tag do %>
+      <% assignable_hash = [@task.assignable].to_json.html_safe %>
```
Comment (zainsaleem4338) [inline app/views/tasks/_form.html.erb]: spacing

File: app/views/tasks/_form.html.erb [hunk 36]
```diff
@@ -0,0 +1,130 @@
...
+    </div>
+    <% else %>
+
+    <label class="col"><input type="radio" name="adding"; value="1" class="js-toggle-team-employee-input-token team_radio" data-type="Employee">Add a task for Employee</label>
+    <label class="col"><input type="radio" name="adding"; value="2" checked="checked" class="js-toggle-team-employee-input-token"  data-type="Team"/>Add a task for Team</label>
+
+    <div class= "adding_employee_field collapse js-toggleable-input-token">
+      <%= form.label :assignable_id, "Employee Email" %><br>
+      <%= text_field_tag :assignable_employee_id, @task.assignable, class: "js-assign-employee" %>
+      <% if @task.assignable_type.nil? %>
```
Comment (zainsaleem4338) [inline app/views/tasks/_form.html.erb]: we should move it to helper method

File: app/views/tasks/_form.html.erb [hunk 37]
```diff
@@ -0,0 +1,130 @@
...
+      <% assignable_hash = [@task.assignable].to_json.html_safe %>
+      $('.js-assign-employee').tokenInput('/employeeslist.json', { prePopulate: <%= assignable_employee_hash %>, tokenLimit: 1 });
+      <% end %>
+    </div>
+
+    <div class= "adding_team_field js-toggleable-input-token">
+      <%= form.label :assignable_id, "Team Name" %><br>
+      <% assignable_hash = [@task.assignable].to_json.html_safe %>
+      <%= text_field_tag :assignable_team_id, @task.assignable, class: "js-assign-team" %>
+      <% if @task.assignable_type.nil? %>
```
Comment (zainsaleem4338) [inline app/views/tasks/_form.html.erb]: we should move it to helper method

File: app/views/tasks/_form.html.erb [hunk 38]
```diff
@@ -0,0 +1,130 @@
...
+      <% assignable_team_hash = nil%>
+      <% elsif @task.assignable_type == "Team" %>
```
Comment (zainsaleem4338) [inline app/views/tasks/_form.html.erb]: constant

File: app/views/tasks/_form.html.erb [hunk 39]
```diff
@@ -0,0 +1,130 @@
...
+      <% assignable_team_hash = [@task.assignable].to_json.html_safe %>
+      <% end %>
+      <%= javascript_tag do %>
+      $('.js-assign-team').tokenInput('/teamslist.json', { prePopulate: <%= assignable_team_hash %>, tokenLimit: 1 });
```
Comment (zainsaleem4338) [inline app/views/tasks/_form.html.erb]: spacing

File: app/views/tasks/index.html.erb [hunk 40]
```diff
@@ -0,0 +1,44 @@
...
+        <th scope="col"></th>
+      </tr>
+    </thead>
+    <tbody>
+      <% @project.tasks.each do |task|%>
+      <tr>
+        <td><%= task.name %></td>
+        <td><%= task.start_date.to_date %></td>
+        <td><%= task.expected_end_date.to_date %></td>
+        <% if task.assignable_type.nil? %>
```
Comment (zainsaleem4338) [inline app/views/tasks/index.html.erb]: move logic to helper method

File: test/controllers/employees_controller_test.rb (hunk 10 above)
Comment (zainsaleem4338) [reply test/controllers/employees_controller_test.rb]: why was this comment not resolved?

File: test/controllers/tasks_controller_test.rb (hunk 12 above)
Comment (zainsaleem4338) [reply test/controllers/tasks_controller_test.rb]: why was this comment not resolved?

File: test/controllers/teams_controller_test.rb (hunk 13 above)
Comment (zainsaleem4338) [reply test/controllers/teams_controller_test.rb]: why was this comment not resolved?

File: app/controllers/tasks_controller.rb (hunk 27 above)
Comment (mabdullahkhalil) [reply app/controllers/tasks_controller.rb]: That is present but these are like hidden ones. The controller does the job based on this.

File: test/models/company_test.rb (hunk 15 above)
Comment (zainsaleem4338) [reply test/models/company_test.rb]: why was this comment not resolved?

File: test/models/employee_test.rb (hunk 17 above)
Comment (zainsaleem4338) [reply test/models/employee_test.rb]: why was this comment not resolved?

File: app/controllers/employees_controller.rb (hunk 26 above)
Comment (mabdullahkhalil) [reply app/controllers/employees_controller.rb]: yes, have changed it. sumbal did it too.

File: test/models/project_test.rb (hunk 18 above)
Comment (zainsaleem4338) [reply test/models/project_test.rb]: why was this comment not resolved?

File: test/models/task_test.rb (hunk 19 above)
Comment (zainsaleem4338) [reply test/models/task_test.rb]: why was this comment not resolved?

File: test/models/team_test.rb (hunk 20 above)
Comment (zainsaleem4338) [reply test/models/team_test.rb]: why was this comment not resolved?

File: app/views/tasks/_form.html.erb (hunk 36 above)
Comment (mabdullahkhalil) [reply app/views/tasks/_form.html.erb]: We will but not now please. :p

**Pull request merged**